federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -681,27 +681,30 @@ public boolean isValid(RealmModel realm, UserModel user, CredentialInput input)
     }
 
     @Override
-    public CredentialValidationOutput authenticate(RealmModel realm, CredentialInput cred) {
+    public CredentialValidationOutput authenticate(RealmModel realm, CredentialInput cred, CredentialValidationOutput prevAttempt) {
         if (!(cred instanceof UserCredentialModel)) CredentialValidationOutput.failed();
         UserCredentialModel credential = (UserCredentialModel)cred;
         if (credential.getType().equals(UserCredentialModel.KERBEROS)) {
             if (kerberosConfig.isAllowKerberosAuthentication()) {
                 String spnegoToken = credential.getValue();
                 SPNEGOAuthenticator spnegoAuthenticator = factory.createSPNEGOAuthenticator(spnegoToken, kerberosConfig);
 
-                spnegoAuthenticator.authenticate();
+                if (!isKerberosUsernameExists(prevAttempt)) { 
```

Comment:
I think that this means that if you have 2 LDAP servers "ldap-1" and "ldap-2" and the SPNEGO_AUTHENTICATED_USER would be added to the state by "ldap-1", then "ldap-2" won't try spnego authentication at all, but just lookup user. This is not correct IMO. The same LDAP server, which did SPNEGO authentication, also needs to lookup user.

Reply:
I did try your suggested implementation at first. But since the first kerberos authentication failed (for ldap-1 as suggested in Bill comment KEYCLOAK-6225) Kerberos will reject further attempts with an error. Since we can't initiate a new kerberos authentication attempt without a client redirect, i don't think we have an alternative solution.

Reply:
Hmm... That's strange. In theory, the call to "gssContext.acceptSecContext" shouldn't communicate with Kerberos server at all, so it is a bit strange that second call won't work. But maybe you're right, I didn't try by myself...